torch/nn/modules/transformer.py

```diff
@@ -314,20 +334,8 @@ def forward(
                 output = torch._nested_tensor_from_mask(output, src_key_padding_mask.logical_not(), mask_check=False)
                 src_key_padding_mask_for_layers = None
 
-        # Prevent type refinement
-        make_causal = (is_causal is True)
-
-        if is_causal is None:
-            if mask is not None:
-                sz = mask.size(0)
-                causal_comparison = torch.triu(
-                    torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
-                ).to(mask.dtype)
-
-                if torch.equal(mask, causal_comparison):
-                    make_causal = True
-
-        is_causal = make_causal
+        seq_len = None if (src.is_nested or not hasattr(self, "seq_len_pos")) else src.size(self.seq_len_pos)
+        is_causal = _detect_is_causal_mask(mask, is_causal, seq_len)
```

Comment:
Do we care about preserving BC for the case where a user supplied a mask that `is_causal` but is of the wrong size?

IMO, this seems like a bug, so we don't need to provide a BC guarantee here (we can just  do a check and error if causal_mask is of the wrong size but lmk your thoughts)

Reply:
The compatibility issue is that on old models, the attribute "seq_len_pos" is not defined.  (There's gotta be a better way, but this is what we have in models all over the place -- newly introduced attributes have to be checked if they have the attribute, because data structures (I guess loaded from pickle files?) have the old definition but the new code.  

I'm happy to remove the hasattr check (actually this was the version I had before, it def looks nicer)

Reply:
OK, rather than using a new field, I compute seq_len_pos every time, so we don't depend on an attribute that may not be present in objects created with a previous version of torch